root.c

```diff
@@ -1922,6 +1938,161 @@ static void notify_thread(struct unlocked_watchman_root *unlocked)
+    w_string_t *dir_name_real_wstr =
+      w_string_new_typed(dir_name_real, W_STRING_BYTE);
+    normalized_target = w_string_path_cat(dir_name_real_wstr, file_name);
+    w_string_delref(dir_name_real_wstr);
+    free(dir_name_real);
+  }
+  w_string_delref(dir_name);
+  w_string_delref(file_name);
+  free(dir_name_buf);
+  return normalized_target;
+}
+
+// Requires target to be an absolute path
+static void watch_symlink_target(w_string_t *target) {
+  char *watched_root = NULL, *relpath = NULL;
+  w_string_t *normalized_target;
+
+  w_assert(w_is_path_absolute(target->buf),
+           "watch_symlink_target: path %s is not absolute\n", target->buf);
+  normalized_target = get_normalized_target(target);
+  if (!normalized_target) {
+    w_log(W_LOG_ERR, "watch_symlink_target: "
+          "failed to get normalized version of target %s\n", target->buf);
+    return;
+  }
+  watched_root = w_find_enclosing_root(normalized_target->buf, &relpath);
+  if (watched_root) {
+    // We are already watching a root that contains this target
+    free(watched_root);
+    if (relpath) {
+      free(relpath);
+    }
+  } else {
+    char *resolved, *errmsg;
+    json_t *root_files;
+    bool enforcing, res;
+    resolved = strdup(normalized_target->buf);
+    root_files = cfg_compute_root_files(&enforcing);
+    if (!root_files) {
+      w_log(W_LOG_ERR,
+            "watch_symlink_target: error computing root_files configuration "
+            "value, consult your log file at %s for more details\n", log_name);
+    } else {
+      res = find_project_root(root_files, resolved, &relpath);
+      if (!res) {
+        w_log(W_LOG_ERR, "No root project found to contain %s\n", resolved);
+      } else {
+        struct unlocked_watchman_root unlocked;
+        bool success = w_root_resolve(resolved, true, &errmsg, &unlocked);
+        if (!success) {
+          w_log(W_LOG_ERR, "watch_symlink_target: failed to watch %s\n",
+                resolved);
+        }
+      }
+      json_decref(root_files);
+    }
+    // Freeing resolved also frees rel_path
+    free(resolved);
+  }
+  w_string_delref(normalized_target);
+}
+
+/** Given an absolute path, watch all symbolic links associated with the path.
+ * Since the target of a symbolic link might contain several components that
+ * are themselves symlinks, this function gets called recursively called on
+ * all the components of path. */
+static void watch_symlinks(w_string_t *path) {
+  w_string_t *dir_name, *file_name;
+  char link_target_path[WATCHMAN_NAME_MAX];
+  ssize_t tlen = 0;
+
+  // We do not currently support symlinks on Windows, so comparing path to "/"
+  // is ok
+  if (!path || w_string_equal_cstring(path, "/")) {
+    return;
+  }
+  w_assert(w_is_path_absolute(path->buf),
+           "watch_symlinks: path %s is not absolute\n", path->buf);
+  dir_name = w_string_dirname(path);
+  file_name = w_string_basename(path);
+  tlen = readlink(path->buf, link_target_path, WATCHMAN_NAME_MAX);
```

Comment:
since `path` may be the result of a `w_string_dirname` call when we recurse (the `dir_name` assignment above), it may be a slice and therefore `path->buf` may not be nul-terminated.

You'll need to call `w_string_dup_buf(path)` to guarantee that you have a nul-terminated string.